presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -367,6 +376,24 @@ public void enforceMemoryLimits()
     }
 
     /**
+     * Enforce timeout at the query level
+     */
+    public void enforceQueryMaxRuntimeLimits()
+    {
+        Iterable<QueryExecution> runningQueries = queries.values().stream()
+                .filter(query -> query.getQueryInfo().getState() == RUNNING)
+                .collect(toImmutableList());
+
+        for (QueryExecution runningQuery : runningQueries) {
+            Duration queryMaxRuntime = SystemSessionProperties.getQueryMaxRuntime(runningQuery.getSession());
```

Comment:
I think we want to fail queries in any non-done state if they have exceed the limit.  For example, if we have a bug in the planner, then we want to fail the query after the time limit.  

Also, note the query state machine will not let you change the state of a query in a done state, so you can simply process all queries and fail those that have exceeded the limit. If any happen to be finished when you fail them, the call will be ignored.